lib/handler.js

```diff
@@ -0,0 +1,101 @@
+const errors = require('./errors');
+const lambdaHandler = require('lambda-handler-as-promised');
+const marshaler = require('dynamodb-marshaler');
+const schemas = require('./schemas');
+const utils = require('./utils');
+
+module.exports = function(options) {
+  options = options || {}; // eslint-disable-line no-param-reassign
+
+  utils.validate(options, schemas.HANDLER_OPTIONS);
+
+  const esclient = aeclient.create(options.elasticsearch || options.es);
+  const separator = _.has(options, 'separator') ? options.separator : '.';
+
+  return lambdaHandler((event, context) => {
+    return Promise.resolve().then(() => {
+      if (options.beforeHook) {
+        options.beforeHook(event, context);
+      }
+
+      utils.validate(event, schemas.EVENT, { allowUnknown: true });
+
+      const actions = [];
+      event.Records.forEach(record => {
+        try {
+          const parsedRecord = marshaler.toJS({
+            newDoc: { M: record.dynamodb.NewImage || {} },
+            oldDoc: { M: record.dynamodb.OldImage || {} },
+            keys: { M: record.dynamodb.Keys }
+          });
+
+          const id = options.idField
+          ? utils.assembleField(parsedRecord, options.idField, separator)
+          : utils.assembleField(parsedRecord, _.keys(parsedRecord.keys), separator);
+
+          const actionDescriptionObj = {
+            _index: options.index
+              || utils.assembleField(parsedRecord, options.indexField, separator),
+            _type: options.type
+              || utils.assembleField(parsedRecord, options.typeField, separator),
+            _id: id
+          };
+          const doc = options.pickFields
+          ? _.pick(parsedRecord.newDoc, options.pickFields)
+          : parsedRecord.newDoc;
+
+          switch (record.eventName) {
+            case 'INSERT':
+              actions.push({ create: actionDescriptionObj });
+              actions.push(doc);
+              break;
+
+            case 'MODIFY':
+              actions.push({ update: actionDescriptionObj });
+              actions.push({ doc });
+              break;
+
+            case 'REMOVE':
+              actions.push({ delete: actionDescriptionObj });
+              break;
+
+            default:
+              throw new errors.UnknownEventNameError(record);
+          }
+        } catch (err) {
+          if (options.recordErrorHook) {
+            options.recordErrorHook(event, context, err);
+            return;
+          }
+          throw err;
+        }
+      });
+
+      if (actions.length === 0) {
+        return {
+          took: 0,
+          errors: false,
+          items: []
+        };
+      }
+
+      return esclient.bulk({ body: actions });
+    })
+    .then(result => {
+      if (options.afterHook) {
+        options.afterHook(event, context, result);
+      }
+      return result;
+    })
+    .catch(err => {
+      if (options.errorHook) {
+        return options.errorHook(event, context, err);
+      }
+      throw err;
```

Comment:
[From some reading](http://stackoverflow.com/questions/33445415/javascript-promises-reject-vs-throw) and my experiments: there is absolutely no difference between two options. Stack is exactly the same.